### PR TITLE
[SYCL][L0] Add support for pinned host memory.

### PR DESCRIPTION
**Pull request by rbegam**
This change implements the support for pinned host memory in level_zero plugin. It allocates pinned host memory when PI_MEM_FLAGS_HOST_PTR_ALLOC is set. Pinned host memories are automatically accessible from the device through PCI. This property also ensures that the memory map/unmap operations are free of cost and the buffer is optimized for frequent accesses from the host.
A test is added for this change in a separate PR for the intel/llvm-test-suite: https://github.com/intel/llvm-test-suite/pull/83

Signed-off-by: rbegam <rehana.begam@intel.com>

### CODE REVIEW

**Review by smaslov-intel — CHANGES_REQUESTED** [review]
see comments

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -1881,8 +1881,24 @@ pi_result piMemBufferCreate(pi_context Context, pi_mem_flags Flags, size_t Size,
                              &Ptr));
   }
 
-  if ((Flags & PI_MEM_FLAGS_HOST_PTR_USE) != 0 ||
-      (Flags & PI_MEM_FLAGS_HOST_PTR_COPY) != 0) {
+  if ((Flags & PI_MEM_FLAGS_HOST_PTR_ALLOC) != 0) {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1884]: add comments what this is doing and explain how this is connected to "pinned" memory

Comment (rbegam) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1884]: I am waiting on a query update. I'll modify it as soon as I have it.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1884]: Is this complete now? Still, can you explain how this is going to be used? Shouldn't the 
implementation use allocated memory as the storage bits for the memory object, similar to PI_MEM_FLAGS_HOST_PTR_USE? 

What is the "pinned" memory extension exactly?

**Review by smaslov-intel — CHANGES_REQUESTED** [review]
Add a test

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1941]: @rbegam ping

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -1935,8 +1935,20 @@ pi_result piMemBufferCreate(pi_context Context, pi_mem_flags Flags, size_t Size,
...
+  // host memory. So even if PI_MEM_FLAGS_HOST_PTR_ALLOC set, it allocates
+  // pageable host memory.
+  if ((Flags & PI_MEM_FLAGS_HOST_PTR_ALLOC) != 0) {
+    ze_host_mem_alloc_desc_t ZeHostDesc = {};
+    ZeHostDesc.flags = 0;
+
+    ZE_CALL(zeMemAllocHost(Context->ZeContext, &ZeHostDesc, Size,
+                           1, // TODO: alignment
+                           &HostPtr));
+    ZE_CALL(zeCommandListAppendMemoryCopy(Context->ZeCommandListInit, Ptr,
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1948]: Please explain why would anyone need the buffer to be allocated in both device and host memory at the same time? And why a copy of just allocated memory (uninitialized) is wanted?

**Review by smaslov-intel — CHANGES_REQUESTED** [review]
see comments

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (rbegam) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1948]: use_pinned_host_memory extension is introduced to support those sycl BEs that can "accelerate copies between host and device by allocating pinned memory". As of now L0 don't do that so this allocation is not going to improve or help in our case. It's my understanding that when PI_MEM_FLAGS_HOST_PTR_ALLOC flag is set we can either 1) report unsupported for it (current) or 2) allocate regular host memory (assuming it wasn't done) and do a device <-> host mapping. Please let me know if you have any suggestion on how to handle it.
Uninitialized mem copy is not necessary I agree.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: > use_pinned_host_memory extension is introduced to support those sycl BEs that can "accelerate copies between host and device by allocating pinned memory"

@rbegam please point me to the extension definition

> L0 don't do that so this allocation is not going to improve or help in our case.

Performance put aside for a moment, how host memory allocation is going to be useful for the pinned memory feature?

> and do a device <-> host mapping (like cuda).

I am missing to see how this "mapping" is performed.

@romanovvlad could you clarify the circumstances at which SYCL RT calls the `piMemBufferCreate` with `PI_MEM_FLAGS_HOST_PTR_ALLOC` and what's the expectation there?

Comment (romanovvlad) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: The extension definition is here: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/UsePinnedMemoryProperty/UsePinnedMemoryPropery.adoc

> @romanovvlad could you clarify the circumstances at which SYCL RT calls the piMemBufferCreate with PI_MEM_FLAGS_HOST_PTR_ALLOC and what's the expectation there?

When a user passes `sycl::ext::oneapi::property::buffer::use_pinned_host_memory` property to the buffer constructor.
I would expect that an implementation allocates a buffer optimized for host<->device copy if the property is passed.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: The only semantic description given so far is this:
> Some SYCL backends can accelerate copies between host and device by allocating pinned memory.

1. What should do the BE that can't accelerate?
2. How (and why) `sycl::ext::oneapi::property::buffer::use_pinned_host_memory` translates into `PI_MEM_FLAGS_HOST_PTR_ALLOC`? The later does not seem to hold the same semantics.

Comment (romanovvlad) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: 1. I think BE should just ignore it, but @jbrodman had a different opinion on that.
2. I tend to agree. This is how it's done currently. Could you please tell if you think we need to introduce a new flag?

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: @jbrodman please share your opinion here?
@romanovvlad , are there other cases where PI_MEM_FLAGS_HOST_PTR_ALLOC is used?

Comment (romanovvlad) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: > @romanovvlad , are there other cases where PI_MEM_FLAGS_HOST_PTR_ALLOC is used?

In the SYCL RT - no.

Comment (rbegam) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: there is a recent commit for buffer management https://github.com/intel/llvm/commit/2ae1bc9e32c3a2dc02add5694d593ad4027d9cc2. For the case where there is an integrated device, this change introduces the similar behavior as pinned memory. 
@romanovvlad is there any correlation between PI_MEM_FLAGS_HOST_PTR_ALLOC and integrated device/ZE_DEVICE_PROPERTY_FLAG_INTEGRATED?

Comment (romanovvlad) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: I think for integrated devices all memory is "fast" since it's shared with host. The pinned memory makes more sense for discrete devices. I'm not an expert though.

Comment (rbegam) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: should I just ignore the flag as @romanovvlad suggested then? @smaslov-intel 
@jbrodman ping

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: I think you need to do what this asks for -- allocates buffer in host-accessible memory (https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clCreateBuffer.html). Note it is not to allocate host memory AND device memory but rather to allocate shared memory for the buffer.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp]: We do need a test for this. Please make sure to cover the interaction of such (pinned) allocated memory and mapping/unmapping.

**Comment by smaslov-intel** [comment]
We do need a test for this. Please make sure to cover the interaction of such (pinned) allocated memory and mapping/unmapping.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 3]
```diff
@@ -1929,20 +1933,32 @@ pi_result piMemBufferCreate(pi_context Context, pi_mem_flags Flags, size_t Size,
...
+  } else if (AllocHostPtr) {
+    // Currently L0 does not support allocation of pinned
+    // host memory. So for PI_MEM_FLAGS_HOST_PTR_ALLOC flag, it allocates
+    // from host accessible memory.
+    ze_host_mem_alloc_desc_t ZeHostMemDesc = {};
+    ZeHostMemDesc.flags = 0;
 
-    ZE_CALL(
-        zeMemAllocDevice(Context->ZeContext, &ZeDesc, Size, 1, ZeDevice, &Ptr));
+    ZE_CALL(zeMemAllocShared(Context->ZeContext, &ZeDeviceMemDesc,
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1951]: If the `DeviceIsIntegrated` then the `zeMemAllocHost` is sufficient, and I believe is faster. Otherwise, as requested by user, the allocation should be accessible from the host, so `zeMemAllocSahred` seems good then. I am not sure what will be the interaction of such allocated buffer with the map/unmap operations, whose implementations are highly coupled with this create one. Please add a functional test.

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 4]
```diff
@@ -1957,6 +1973,8 @@ pi_result piMemBufferCreate(pi_context Context, pi_mem_flags Flags, size_t Size,
                                               HostPtr, Size, nullptr, 0,
                                               nullptr));
       }
+    } else if ((Flags & PI_MEM_FLAGS_HOST_PTR_ALLOC) != 0) {
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1976]: You are not setting the `HostPtr` to the allocated memory, so you can't get here, right? (line 1962 should be false). So why is this change needed?

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1976]: Maybe instead verify that `HostPtr` is null when PI_MEM_FLAGS_HOST_PTR_ALLOC is used?

**Review by smaslov-intel — CHANGES_REQUESTED** [review]

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 4 above)
Comment (rbegam) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1976]: @smaslov-intel are you saying that hostPtr and PI_MEM_FLAGS_HOST_PTR_ALLOC is mutually exclusive? My understanding was that they are not.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 3 above)
Comment (rbegam) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1951]: there is a test for pinned memory https://github.com/intel/llvm/pull/2080/files?file-filters%5B%5D=.cpp#diff-c0fd7ed69152328e50c4747dc8870b42d5e085b05a3f20cba5c42f79a4002e14

**Comment by jbrodman** [comment]
I'm confused what this PR is doing.

Is it adding support for https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions/UsePinnedMemoryProperty?

If so I'm confused: I don't get why USM shared allocations are used here at all.
It seems like a buffer should be represented as 2 allocations: 1 in host memory, 1 in device memory.
The property says that the host memory one should likely be done as a USM host allocation instead of system malloc as that usually means it's pinned.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 4 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1976]: Yes, see in https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clCreateBuffer.html:

> CL_​MEM_​ALLOC_​HOST_​PTR and CL_​MEM_​USE_​HOST_​PTR are mutually exclusive.

PI_MEM_FLAGS_HOST_PTR_ALLOC is analogous to CL_​MEM_​ALLOC_​HOST_​PTR

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 5]
```diff
@@ -2002,6 +2002,20 @@ pi_result piMemBufferCreate(pi_context Context, pi_mem_flags Flags, size_t Size,
...
+
+  if (AllocHostPtr) {
+    assert(HostPtr == nullptr &&
+           "PI_MEM_FLAGS_HOST_PTR_ALLOC cannot be used with host pointer");
+
+    ze_host_mem_alloc_desc_t ZeDesc = {};
+    ZeDesc.flags = 0;
+
+    ZE_CALL(zeMemAllocHost(Context->ZeContext, &ZeDesc, Size, 1, &Ptr));
+  }
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2017]: This can't be right. The code below will do another memory allocation for `Ptr` effectively leaking memory allocated in line 2016. Given the fact that there is no test failed, I am asking again that a test be added, which checks the desired behavior intended to be added here.

Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2017]: @rbegam ping

Comment (rbegam) [reply sycl/plugins/level_zero/pi_level_zero.cpp:2017]: @smaslov-intel I was under the impression that these two cases are mutually exclusive. I'll follow up on our discussion and update it.

File: sycl/plugins/level_zero/pi_level_zero.hpp [hunk 6]
```diff
@@ -390,6 +390,7 @@ struct _pi_mem : _pi_object {
   char *MapHostPtr;
 
   // Flag to indicate that this memory is allocated in host memory
+  // if created with PI_MEM_FLAGS_HOST_PTR_ALLOC and/or integrated devices.
```
Comment (smaslov-intel) [inline sycl/plugins/level_zero/pi_level_zero.hpp:393]: NIT: Let's not list all the cases when memory is allocated on host, since that might change in future.

**Review by smaslov-intel — DISMISSED** [review]
LGTM

File: sycl/test/on-device/basic_tests/use_pinned_host_memory.cpp [hunk 7]
```diff
@@ -22,6 +32,27 @@ int main() {
     CGH.single_task<class init_a>([=]() {});
   });
 
+  {
+    int data1[10] = {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1};
+    {
+      buffer<int, 1> a(data1, range<1>(10), {property::buffer::use_host_ptr()});
+      buffer<int, 1> b(
+          range<1>(10),
+          {ext::oneapi::property::buffer::use_pinned_host_memory()});
```
Comment (romanovvlad) [inline sycl/test/on-device/basic_tests/use_pinned_host_memory.cpp:41]: Can we have a test which verifies that piMemBufferCreate with PI_MEM_FLAGS_HOST_PTR_ALLOC allocates memory using zeMemAllocHost?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 8]
```diff
@@ -2088,7 +2088,24 @@ pi_result piMemBufferCreate(pi_context Context, pi_mem_flags Flags, size_t Size,
...
+    PI_ASSERT(HostPtr == nullptr, PI_INVALID_VALUE);
+
+    ze_host_mem_alloc_desc_t ZeDesc = {};
+    ZeDesc.flags = 0;
+
+    ZE_CALL(zeMemAllocHost(Context->ZeContext, &ZeDesc, Size, 1, &Ptr));
+
+  } else if (DeviceIsIntegrated) {
     ze_host_mem_alloc_desc_t ZeDesc = {};
     ZeDesc.flags = 0;
```
Comment (romanovvlad) [inline sycl/plugins/level_zero/pi_level_zero.cpp:2111]: Minor. Probably, these two branches could be merged.

File: sycl/test/on-device/basic_tests/use_pinned_host_memory.cpp [hunk 9]
```diff
@@ -1,10 +1,20 @@
...
+// RUN: %clangxx -fsycl -fsycl-targets=%sycl_triple %s -o %t2.out
+// RUN: env SYCL_PI_TRACE=2 %CPU_RUN_PLACEHOLDER %t2.out 2>&1 %CPU_CHECK_PLACEHOLDER
+
+// RUN: %RUN_ON_HOST %t2.out
+// RUN: %CPU_RUN_PLACEHOLDER %t2.out
+// RUN: %GPU_RUN_PLACEHOLDER %t2.out
+// RUN: %ACC_RUN_PLACEHOLDER %t2.out
 
 #include <CL/sycl.hpp>
+#include <cassert>
```
Comment (romanovvlad) [inline sycl/test/on-device/basic_tests/use_pinned_host_memory.cpp:13]:
```suggestion

#include <cassert>
```
Minor.

**Comment by rbegam** [comment]
the use_pinned_host_memory.cpp test seems to be moved into intel/llvm-test-suite.
@romanovvlad I am going to remove the test from this PR and submit a new one for intel/llvm-test-suite. is that okay?

**Comment by romanovvlad** [comment]
> 
> 
> the use_pinned_host_memory.cpp test seems to be moved into intel/llvm-test-suite.
> @romanovvlad I am going to remove the test from this PR and submit a new one for intel/llvm-test-suite. is that okay?

That's OK.

**Comment by rbegam** [comment]
ping @smaslov-intel

**Review by smaslov-intel — APPROVED** [review]
LGTM. It would be great to put a link to the test in the PR (or better yet into description of the commit).

**Comment by pvchupin** [comment]
@rbegam, do you have link to test? This change seems ready to merge.

**Comment by rbegam** [comment]
@pvchupin yes, I am updating the commit msg with it.
